From ed177ccb9c8f5192baab71fed683b851cb616ff8 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Date: Fri, 12 Jul 2013 10:32:13 +0200
Subject: [PORT FROM MAIN]ueventd: Overload the do_setprop builtin

BZ: 125765

This patch overloads the do_setprop builtin so we can call it from another
process than init, particularly from ueventd. This extra function makes sure to
keep the properties in a temporary area if the init process hasn't initialized
the property_service at the time the property needs to be defined. This is
required so we can later on add ueventd actions which will set properties on
receival of certain modalias, which can happen really early in the boot
process.

Change-Id: I18a0d5c3d5d9a82087434eeb6f4b95de734290be
Orig-Change-Id: I56f74fb2d4a0f164e329ba2ddff21350c8363ca8
Category: device enablement
Domain: CWS.WIFI-Common
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Reviewed-on: http://android.intel.com:8080/118419
Reviewed-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Signed-off-by: Jeremie GARCIA <jeremie.garcia@intel.com>
---
 init/builtins.c         |   67 +++++++++++++++++++++++++++++++++++++++++++++++
 init/keywords.h         |    3 +-
 init/property_service.c |    1 +
 init/property_service.h |    2 +
 4 files changed, 72 insertions(+), 1 deletions(-)

diff --git a/init/builtins.c b/init/builtins.c
index b9fff40..be8f91e 100644
--- a/init/builtins.c
+++ b/init/builtins.c
@@ -59,6 +59,14 @@ enum builtin_cmds {
 
 #define MAX_RECUR_DEPTH 15
 
+#ifndef PROPERTY_VALUE_MAX
+#  define PROPERTY_VALUE_MAX 92
+#endif
+
+#ifndef PROPERTY_KEY_MAX
+#  define PROPERTY_KEY_MAX 32
+#endif
+
 void add_environment(const char *name, const char *value);
 
 extern int init_module(void *, unsigned long, const char *);
@@ -785,6 +793,65 @@ int do_setprop(int nargs, char **args)
     return 0;
 }
 
+static int append_to_file(const char* path, const char *data, size_t len)
+{
+    int fd = -1;
+    int ret = 0;
+
+    fd = TEMP_FAILURE_RETRY(open(path, O_WRONLY | O_CREAT | O_APPEND, 0640));
+    if (fd < 0) {
+        ERROR("Failed to open %s (%s)",
+              path, strerror(errno));
+        return -errno;
+    }
+
+    if (TEMP_FAILURE_RETRY(write(fd, data, len)) != (int) len) {
+        ERROR("Failed to write %s in %s (%s)",
+              data, path, strerror(errno));
+        ret = -errno;
+    }
+
+    close(fd);
+    return ret;
+}
+
+static int uevent_temporary_setprop(char* name, char* prop_val)
+{
+    /*  "key" + "=" + "value" + "\n\0" */
+    char buff[PROPERTY_KEY_MAX + 1 + PROPERTY_VALUE_MAX + 2];
+    int len = -1;
+
+    len = snprintf(buff, sizeof(buff), "%s=%s\n", name, prop_val);
+
+    append_to_file(PROP_PATH_UEVENTD, buff, (len >= (int) sizeof(buff) ? sizeof(buff) : len));
+
+    return 0;
+}
+
+int do_ext_setprop(int nargs, char **args)
+{
+    const char *name = args[1];
+    const char *value = args[2];
+    char *prop_val;
+    int ret;
+
+    prop_val = expand_references(value);
+    if (!prop_val) {
+        ERROR("cannot expand '%s' while assigning to '%s'\n", value, name);
+        return -EINVAL;
+    }
+    if ((ret = __system_property_set(name, prop_val)) < 0) {
+        /*
+         * property_service is not running at this time so we put the
+         * properties in temporary space.
+         */
+        ret = uevent_temporary_setprop(name, prop_val);
+    }
+    free(prop_val);
+
+    return ret;
+}
+
 int do_setrlimit(int nargs, char **args)
 {
     struct rlimit limit;
diff --git a/init/keywords.h b/init/keywords.h
index 72d1c56..a2abfd0 100644
--- a/init/keywords.h
+++ b/init/keywords.h
@@ -26,6 +26,7 @@ int do_setcon(int nargs, char **args);
 int do_setenforce(int nargs, char **args);
 int do_setkey(int nargs, char **args);
 int do_setprop(int nargs, char **args);
+int do_ext_setprop(int nargs, char **args);
 int do_setrlimit(int nargs, char **args);
 int do_setsebool(int nargs, char **args);
 int do_start(int nargs, char **args);
@@ -85,7 +86,7 @@ enum {
     KEYWORD(setenforce,  COMMAND, 1, do_setenforce, 0)
     KEYWORD(setenv,      OPTION,  2, 0, 0)
     KEYWORD(setkey,      COMMAND, 0, do_setkey, 0)
-    KEYWORD(setprop,     COMMAND, 2, do_setprop, 0)
+    KEYWORD(setprop,     COMMAND, 2, do_setprop, do_ext_setprop)
     KEYWORD(setrlimit,   COMMAND, 3, do_setrlimit, 0)
     KEYWORD(setsebool,   COMMAND, 1, do_setsebool, 0)
     KEYWORD(socket,      OPTION,  0, 0, 0)
diff --git a/init/property_service.c b/init/property_service.c
index 61e14c7..e52dd33 100644
--- a/init/property_service.c
+++ b/init/property_service.c
@@ -655,6 +655,7 @@ void start_property_service(void)
 
     listen(fd, 8);
     property_set_fd = fd;
+    load_properties_from_file(PROP_PATH_UEVENTD, NULL);
 }
 
 int get_property_set_fd()
diff --git a/init/property_service.h b/init/property_service.h
index 46cbd8f..4deed28 100644
--- a/init/property_service.h
+++ b/init/property_service.h
@@ -20,6 +20,8 @@
 #include <stdbool.h>
 #include <sys/system_properties.h>
 
+#define PROP_PATH_UEVENTD "/dev/.ueventd_properties"
+
 extern void handle_property_set_fd(void);
 extern void property_init(void);
 extern void property_load_boot_defaults(void);
-- 
1.7.4.1

